File: src/langnostic/feed.clj
(ns langnostic.feed
  (:require [hiccup.core :as hic]
            [hiccup.page :as pg]

            [langnostic.page :as page]))

(defn atom-feed [posts]
  (str "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
       (hic/html
        [:feed {:xml:lang "en-US" :xmlns "http://www.w3.org/2005/Atom"}
         [:title "Language Agnostic"]
         [:subtitle "Langnostic Atom Feed"]
         [:link {:href "http://langnostic.inaimathi.ca/feed/atom" :rel "self"}]
         [:link {:href "http://langnostic.inaimathi.ca"}]
         (map (fn [post]
                [:entry
                 [:title (post :title)]
                 [:updated (post :posted)]
                 [:link {:href (str "http://langnostic.inaimathi.ca/posts/" (post :file))}]
                 [:author [:name "inaimathi"]]
                 [:content {:type "html"} (hic/h (post :content))]])
              (take 10 (reverse (sort-by :posted posts))))])))

File: src/langnostic/feed.py
from typing import Any, Dict, List
from xml.sax.saxutils import escape

import pyccup.core

from . import posts


def atom_feed(post_list: List[Dict[str, Any]]) -> str:
    """Generate an Atom feed from the given posts."""

    def format_entry(post):
        return [
            "entry",
            ["title", post.get("title", "")],
            ["updated", post.get("posted", "")],
            [
                "link",
                {
                    "href": f"http://langnostic.inaimathi.ca/posts/{post.get('file', '')}"
                },
            ],
            ["author", ["name", "inaimathi"]],
            ["content", {"type": "html"}, escape(posts.post_content(post))],
        ]

    # Take the latest 20 posts, sorted by posted date in reverse
    sorted_posts = sorted(post_list, key=lambda p: p.get("posted", ""), reverse=True)[
        :20
    ]
    entries = [format_entry(post) for post in sorted_posts]

    feed_tree = [
        ["title", "Language Agnostic"],
        ["subtitle", "Langnostic Atom Feed"],
        ["link", {"href": "http://langnostic.inaimathi.ca/feed/atom", "rel": "self"}],
        ["link", {"href": "http://langnostic.inaimathi.ca"}],
    ]
    feed_tree.extend(entries)

    return pyccup.core.xml(feed_tree, "feed")
